ArduPlane/soaring.cpp

```diff
@@ -47,175 +47,134 @@ void Plane::update_soaring() {
                     break;
                 case SoaringController::LoiterStatus::ALT_TOO_LOW:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooLow, mode_auto.name());
-                    set_mode(mode_auto, ModeReason::SOARING_ALT_TOO_LOW);
+                    gcs().send_text(MAV_SEVERITY_INFO, "Soaring: Too low, restoring %s", exit_mode->name());
+                    set_mode(*exit_mode, ModeReason::SOARING_ALT_TOO_LOW);
                     break;
                 default:
                 case SoaringController::LoiterStatus::THERMAL_WEAK:
-                        gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooWeak, mode_auto.name());
-                        set_mode(mode_auto, ModeReason::SOARING_THERMAL_ESTIMATE_DETERIORATED);
+                    gcs().send_text(MAV_SEVERITY_INFO, "Soaring: Thermal ended, restoring %s", exit_mode->name());
+                    set_mode(*exit_mode, ModeReason::SOARING_THERMAL_ESTIMATE_DETERIORATED);
                     break;
                 case SoaringController::LoiterStatus::DRIFT_EXCEEDED:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooFar, mode_auto.name());
-                    set_mode(mode_auto, ModeReason::SOARING_DRIFT_EXCEEDED);
+                    gcs().send_text(MAV_SEVERITY_INFO, "Soaring: Drifted too far, restoring %s", exit_mode->name());
+                    set_mode(*exit_mode, ModeReason::SOARING_DRIFT_EXCEEDED);
                     break;
-                } // switch loiterStatus
-            break;
-            } // case AUTO
-        default: // all other modes
+            } // switch loiterStatus
+
             break;
-        } // switch previous_mode
-        break;
+           
         } // case loiter
     } // switch control_mode
+}
 
 
-    if (control_mode == &mode_loiter) {
-        // still in thermal - need to update the wp location
-        g2.soaring_controller.get_target(next_WP_loc);
-
+bool Plane::soaring_exit_heading_aligned() const
+{
+    // Return true if the current heading is aligned with the next objective.
+    // If home is not set, or heading not locked, return true to avoid delaying mode change.
+    switch (previous_mode->mode_number()) {
+        case Mode::Number::AUTO: {
+            //Get the lat/lon of next Nav waypoint after this one:
+            AP_Mission::Mission_Command current_nav_cmd = mission.get_current_nav_cmd();;
+            return plane.mode_loiter.isHeadingLinedUp(next_WP_loc, current_nav_cmd.content.location);
+        }
+        case Mode::Number::FLY_BY_WIRE_B:
+            return (!AP::ahrs().home_is_set() || plane.mode_loiter.isHeadingLinedUp(next_WP_loc, AP::ahrs().get_home()));
+        case Mode::Number::CRUISE:
+            return (!cruise_state.locked_heading || plane.mode_loiter.isHeadingLinedUp_cd(cruise_state.locked_heading_cd));
+        default:
+            break;
     }
-
+    return true;
 }
 
-#endif // SOARING_ENABLED
+#endif // SOARING_ENABLED
```

Comment:
we've lost the EOL

Reply:
Thanks, fixed in https://github.com/ArduPilot/ardupilot/pull/14830